pkg/bootstrapper/service.go

```diff
@@ -0,0 +1,258 @@
+	// Arbitrary wait time
+	timeout := time.Now().Add(serviceWaitTime)
+	for status.State != desiredState {
+		if timeout.Before(time.Now()) {
+			return fmt.Errorf("timeout waiting for service to go to state=%d", desiredState)
+		}
+		time.Sleep(300 * time.Millisecond)
+		status, err = serviceObj.Query()
+		if err != nil {
+			return fmt.Errorf("could not retrieve service status: %v", err)
+		}
+	}
+	return nil
+}
+
+// stopKubeletService stops the Windows service via the Windows service API
+func (service *kubeletService) stopKubeletService() error {
+	isServiceRunning, err := service.isKubeletServiceRunning()
+	if err != nil {
+		return fmt.Errorf("unable to check if kubelet service is running: %v", err)
+	}
+	if !isServiceRunning {
+		return nil
+	}
+
+	if len(service.svcDependents) != 0 {
+		for _, dependent := range service.svcDependents {
+			if err := stopDependentService(dependent); err != nil {
+				return fmt.Errorf("failed to stop services dependent on %s", KubeletServiceName)
+			}
+		}
+	}
+
+	if err := service.controlService(svc.Stop, svc.Stopped); err != nil {
+		return fmt.Errorf("unable to stop Windows Service %s", KubeletServiceName)
+	}
+
+	return nil
+}
+
+// stopDependentService stops the kubelet dependent service
+func stopDependentService(serviceObj *mgr.Service) error {
+	isServiceRunning, err := isDependentRunning(serviceObj)
+	if err != nil {
+		return fmt.Errorf("unable to check if service is running: %v", err)
+	}
+	if isServiceRunning {
+		err := controlDependentService(serviceObj, svc.Stop, svc.Stopped)
+		if err != nil {
+			return fmt.Errorf("unable to stop %s dependent service", serviceObj.Name)
+		}
+	}
+	return nil
+}
+
+// refreshKubeletService updates the kubelet service with the given config and restarts the service
+func (service *kubeletService) refreshKubeletService(config mgr.Config) error {
+	if err := service.stopKubeletService(); err != nil {
+		return fmt.Errorf("error stopping kubelet service: %v", err)
+	}
+
+	if err := service.obj.UpdateConfig(config); err != nil {
+		return fmt.Errorf("error updating kubelet service: %v", err)
+	}
+
+	if err := service.startKubeletService(); err != nil {
+		return fmt.Errorf("error starting kubelet service: %v", err)
+	}
+
+	return nil
+}
+
+// removeKubeletService deletes the kubelet service via the Windows service API
+func (service *kubeletService) removeKubeletService() error {
+	if service.obj == nil {
+		return nil
+	}
+	return service.obj.Delete()
+}
+
+// isKubeletServiceRunning returns true if the kubelet service is running
+func (service *kubeletService) isKubeletServiceRunning() (bool, error) {
+	status, err := service.obj.Query()
+	if err != nil {
+		return false, err
+	}
+	return status.State == svc.Running, nil
+}
+
+// isDependentRunning returns true if the kubelet dependent service is running
+func isDependentRunning(serviceObj *mgr.Service) (bool, error) {
```

Comment:
This mentions kubelet but seems generic

Reply:
Does this function actually check the dependants, or does it just check the status of whats passed into it?
If the latter, this can be renamed to `isRunning`.